invoice/invoice.go

```
@@ -0,0 +1,966 @@
+	// data.
+	hash := chainhash.HashB(toSign)
+
+	// If pubkey was provided as a tagged field, use that to verify the
+	// signature, if not do public key recovery.
+	if decodedInvoice.PubKey != nil {
+		var signature *btcec.Signature
+		lnwire.DeserializeSigFromWire(&signature, sigBytes)
+		if !signature.Verify(hash, decodedInvoice.PubKey) {
+			return nil, fmt.Errorf("invalid invoice signature")
+		}
+	} else {
+		headerByte := recoveryID + 27 + 4
+		compactSign := append([]byte{headerByte}, sigBytes[:]...)
+		pubkey, _, err := btcec.RecoverCompact(btcec.S256(),
+			compactSign, hash)
+		if err != nil {
+			return nil, err
+		}
+		decodedInvoice.RecoveredPubKey = pubkey
+	}
+
+	// Now that we have created the invoice, make sure it has the required
+	// fields set.
+	if err := validateInvoice(&decodedInvoice); err != nil {
+		return nil, err
+	}
+
+	return &decodedInvoice, nil
+}
+
+// Encode takes the given MessageSigner and returns a string encoding this
+// invoice signed by the node key of the signer.
+func (invoice *Invoice) Encode(signer MessageSigner) (string, error) {
+	// First check that this invoice is valid before starting the encoding.
+	if err := validateInvoice(invoice); err != nil {
+		return "", err
+	}
+
+	// The buffer will encoded the invoice data using 5-bit groups (base32).
+	var bufferBase32 bytes.Buffer
+
+	// The timestamp will be encoded using 35 bits, in base32.
+	timestampBase32 := uint64ToBase32(invoice.Timestamp)
+
+	// The timestamp must be exactly 35 bits, which means 7 groups. If it
+	// can fit into fewer groups we add leading zero groups, if it is too
+	// big we fail early, as there is not possible to encode it.
+	if len(timestampBase32) > 7 {
+		return "", fmt.Errorf("timestamp too big: %d", invoice.Timestamp)
+	}
+
+	// Add zero bytes to the first 7-len(timestampBase32) groups, then add
+	// the non-zero groups.
+	zeroes := make([]byte, 7-len(timestampBase32), 7-len(timestampBase32))
+	bufferBase32.Write(zeroes)
```

Ignoring errors here with these two writes.